src/core/connection_pool.c

```diff
@@ -0,0 +1,651 @@
+            (Config->NumberOfConnections / RssProcessorCount);
+
+    //
+    // Begin creating and starting connections.
+    //
+    for (uint32_t i = 0; i < Config->NumberOfConnections; i++) {
+        const uint32_t MaxCreationRetries = RssProcessorCount * MAX_CONNECTION_POOL_RETRY_MULTIPLIER;
+        uint32_t RetryCount = 0;
+
+        for (; RetryCount < MaxCreationRetries; RetryCount++) {
+
+            //
+            // The connection takes ownership of the ServerName parameter, so we must
+            // allocate a copy of it for each connection (attempt).
+            //
+            if (ServerNameCopy == NULL) {
+                ServerNameCopy =
+                    QuicConnPoolAllocServerNameCopy(
+                        Config->ServerName,
+                        ServerNameLength);
+                if (ServerNameCopy == NULL) {
+                    Status = QUIC_STATUS_OUT_OF_MEMORY;
+                    goto Error;
+                }
+            }
+
+            uint32_t NewPort = QuicAddrGetPort(&LocalAddress) + 1;
+            if (NewPort > QUIC_ADDR_EPHEMERAL_PORT_MAX) {
+                NewPort = QUIC_ADDR_EPHEMERAL_PORT_MIN;
+            }
+            QuicAddrSetPort(&LocalAddress, (uint16_t)NewPort);
+
+            QUIC_CONN_POOL_RSS_PROC_INFO* CurrentProc =
+                QuicConnPoolGetRssProcForTuple(
+                    &ToeplitzHash,
+                    &ResolvedRemoteAddress,
+                    &LocalAddress,
+                    RssProcessors,
+                    RssProcessorCount,
+                    RssConfig->RssIndirectionTable,
+                    RssConfig->RssIndirectionTableCount);
+
+            if (CurrentProc->ConnectionCount >= ConnectionsPerProc) {
+                //
+                // This processor already has enough connections on it, so try another port number.
+                //
+                continue;
+            }
+
+            uint16_t PartitionIndex =
+                QuicLibraryGetPartitionFromProcessorIndex(CurrentProc->ProcIndex);
+
+            Status =
+                QuicConnPoolTryCreateConnection(
+                    (QUIC_REGISTRATION*)Config->Registration,
+                    (QUIC_CONFIGURATION*)Config->Configuration,
+                    &PartitionIndex,
+                    Config->Handler,
+                    Config->Context ? Config->Context[i] : NULL,
+                    &ResolvedRemoteAddress,
+                    &LocalAddress,
+                    ServerNameCopy,
+                    Config->ServerPort,
+                    Config->Family,
+                    Config->CibirIdLength,
+                    Config->CibirIds ? Config->CibirIds[i] : NULL,
+                    &Connections[i]);
+
+            //
+            // The connection either owns the ServerNameCopy, or it was freed.
+            //
+            ServerNameCopy = NULL;
+            if (Status == QUIC_STATUS_OUT_OF_MEMORY) {
```

Comment:
Personally, I wouldn't special case out of memory errors. For one, they may be transient, and two there might be many errors you _could_ special case, but it unnecessarily complicates the logic for such rare edge cases.

Reply:
Are memory errors not fatal? Should we retry when there's no memory? Most programs respond to OUT_OF_MEMORY by reducing allocations, when they don't just crash outright.

Reply:
I prefer the simplicity to the special case as well